lib/vrl/stdlib/src/encode_logfmt.rs

```diff
@@ -0,0 +1,305 @@
+) -> Box<dyn Iterator<Item = (String, Value)> + 'a> {
+    let iter = input.into_iter().map(move |(key, value)| {
+        let new_key = if depth > 0 {
+            format!("{}{}{}", parent_key, separator, key)
+        } else {
+            key
+        };
+
+        match value {
+            Object(map) => flatten(map, new_key, separator, depth + 1),
+            Array(array) => {
+                let array_map: BTreeMap<_, _> = array
+                    .into_iter()
+                    .enumerate()
+                    .map(|(key, value)| (key.to_string(), value))
+                    .collect();
+                flatten(array_map, new_key, separator, depth + 1)
+            }
+            _ => Box::new(std::iter::once((new_key, value)))
+                as Box<dyn Iterator<Item = (std::string::String, vrl::Value)>>,
+        }
+    });
+
+    Box::new(iter.flatten())
+}
+
+fn encode_field(output: &mut String, key: &str, value: &Value) {
+    encode_string(output, key);
+    output.write_char('=').unwrap();
+    encode_value(output, value)
+}
+
+pub fn encode(input: BTreeMap<String, Value>, fields: &[String]) -> String {
+    let mut output = String::new();
+
+    let mut input: BTreeMap<_, _> = flatten(input, String::from(""), '.', 0).collect();
+
+    for field in fields.iter() {
+        if let Some(val) = input.remove(field) {
+            encode_field(&mut output, field, &val);
+            output.write_char(' ').unwrap();
+        }
+    }
+
+    for (key, value) in input.iter() {
+        encode_field(&mut output, key, value);
+        output.write_char(' ').unwrap();
+    }
+
+    if output.ends_with(" ") {
+        output.truncate(output.len() - 1)
+    }
+
+    output
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use shared::btreemap;
+
+    test_function![
+        encode_logfmt  => EncodeLogfmt;
+
+        single_element {
+            args: func_args![value:
+                btreemap! {
+                    "lvl" => "info"
+                }
+            ],
+            want: Ok("lvl=info"),
+            tdef: TypeDef::new().bytes().fallible(),
+        }
+
+        multiple_elements {
+            args: func_args![value:
+                btreemap! {
+                    "lvl" => "info",
+                    "log_id" => 12345
+                }
+            ],
+            want: Ok("log_id=12345 lvl=info"),
+            tdef: TypeDef::new().bytes().fallible(),
+        }
+
+        string_with_spaces {
+            args: func_args![value:
+                btreemap! {
+                    "lvl" => "info",
+                    "msg" => "This is a log message"
+                }],
+            want: Ok(r#"lvl=info msg="This is a log message""#),
+            tdef: TypeDef::new().bytes().fallible(),
+        }
+
+        string_with_characters_to_escape {
+            args: func_args![value:
+                btreemap! {
+                    "lvl" => "info",
+                    "msg" => r#"payload: {"code": 200}\n"#
```

Comment:
Can we add another field to test here to test the other escapes and also test that keys are being escaped:

```rust
"another_field" => "some\nfield\\and things"
```

Should be:

```rust
want: Ok(r#"lvl=info msg="payload: {\"code\": 200}\\n" "another field": "some\\nfield\\and things""#),
```

Reply:
👍 added in b7aa5c7